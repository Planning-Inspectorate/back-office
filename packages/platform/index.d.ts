import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
