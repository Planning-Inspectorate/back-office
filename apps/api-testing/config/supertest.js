// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
