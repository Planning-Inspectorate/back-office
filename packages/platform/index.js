export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
