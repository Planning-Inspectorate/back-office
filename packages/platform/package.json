{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
