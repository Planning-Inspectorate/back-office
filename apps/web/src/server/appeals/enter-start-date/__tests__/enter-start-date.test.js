import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { createTestEnvironment } from '../../../../../testing/index.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';
const pageUrl = '/enter-start-date';
const appealIdThatExists = 1;
const appealIdThatDoesNotExist = 0;

describe('appeal-details', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /enter-start-date', () => {
		it('should render the enter start date page if an appeal with a matching appealId exists', async () => {
			const response = await request.get(`${baseUrl}/${appealIdThatExists}${pageUrl}`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});
		it('should render the 404 error page if an appeal with a matching appealId does not exist', async () => {
			const response = await request.get(`${baseUrl}/${appealIdThatDoesNotExist}${pageUrl}`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});
	});

	describe('POST /enter-start-date', () => {
		// TODO: BOAT-105 - update once service is wired up
		// it('should call the service method to set the appeal start date if no errors are present', async () => {
		//
		// });

		it('should re-render the enter start date page with the expected error message if the entered day is empty', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered day is not an integer', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': 'a',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered day is longer than 2 characters', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '111',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered day is less than 1', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '0',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered day is greater than 31', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '32',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered month is empty', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered month is not an integer', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': 'a',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered month is longer than 2 characters', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '111',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered month is less than 1', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '0',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered month is greater than 12', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '13',
				'start-date-year': '2023'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered year is not an integer', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '1',
				'start-date-year': 'aaaa'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered year is empty', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '1',
				'start-date-year': ''
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered year is shorter than 4 characters', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '1',
				'start-date-year': '202'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should re-render the enter start date page with the expected error message if the entered year is longer than 4 characters', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '1',
				'start-date-year': '20233'
			});

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should redirect to the start-date-entered page if no errors are present and an appeal with a matching appealId exists and the service method to set the appeal start date was called successfully', async () => {
			const response = await request.post(`${baseUrl}/${appealIdThatExists}${pageUrl}`).send({
				'start-date-day': '1',
				'start-date-month': '1',
				'start-date-year': '2023'
			});

			expect(response.statusCode).toBe(302);
			expect(response.text).toEqual(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-date-entered'
			);
		});
	});
});
