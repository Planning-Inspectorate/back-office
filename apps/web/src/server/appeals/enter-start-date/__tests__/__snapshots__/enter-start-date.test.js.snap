// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appeal-details GET /enter-start-date should render the enter start date page 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-grid-row\\">
                <div class=\\"govuk-grid-column-full\\"><span class=\\"govuk-caption-l govuk-!-margin-bottom-3\\">351062</span>
                </div>
            </div>
            <form method=\\"post\\" novalidate=\\"novalidate\\">
                <div class=\\"govuk-form-group\\">
                    <fieldset class=\\"govuk-fieldset\\" role=\\"group\\" aria-describedby=\\"start-date-hint\\">
                        <legend class=\\"govuk-fieldset__legend govuk-fieldset__legend--l\\">
                            <h1 class=\\"govuk-fieldset__heading\\"> Enter the case start date</h1>
                        </legend>
                        <div id=\\"start-date-hint\\" class=\\"govuk-hint\\">For example, 27 3 2023</div>
                        <div class=\\"govuk-date-input\\" id=\\"start-date\\">
                            <div class=\\"govuk-date-input__item\\">
                                <div class=\\"govuk-form-group\\">
                                    <label class=\\"govuk-label govuk-date-input__label\\" for=\\"start-date-day\\">Day</label>
                                    <input class=\\"govuk-input govuk-date-input__input govuk-input--width-2\\"
                                    id=\\"start-date-day\\" name=\\"start-date-day\\" type=\\"text\\" inputmode=\\"numeric\\">
                                </div>
                            </div>
                            <div class=\\"govuk-date-input__item\\">
                                <div class=\\"govuk-form-group\\">
                                    <label class=\\"govuk-label govuk-date-input__label\\" for=\\"start-date-month\\">Month</label>
                                    <input class=\\"govuk-input govuk-date-input__input govuk-input--width-2\\"
                                    id=\\"start-date-month\\" name=\\"start-date-month\\" type=\\"text\\" inputmode=\\"numeric\\">
                                </div>
                            </div>
                            <div class=\\"govuk-date-input__item\\">
                                <div class=\\"govuk-form-group\\">
                                    <label class=\\"govuk-label govuk-date-input__label\\" for=\\"start-date-year\\">Year</label>
                                    <input class=\\"govuk-input govuk-date-input__input govuk-input--width-4\\"
                                    id=\\"start-date-year\\" name=\\"start-date-year\\" type=\\"text\\" inputmode=\\"numeric\\">
                                </div>
                            </div>
                        </div>
                    </fieldset>
                </div>
                <div class=\\"govuk-inset-text govuk-!-margin-top-0\\">Entering the start date will create the case timetable and publish the
                    appeal on the Appeals Casework Portal</div>
                <button type=\\"submit\\" class=\\"govuk-button\\"
                data-module=\\"govuk-button\\">Save and continue</button>
            </form>
        </div>
    </div>
</main>"
`;
